instrumentation/pulsar/pulsar-2.8/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/pulsar/v2_8/AbstractPulsarClientTest.java

```diff
@@ -0,0 +1,322 @@
+    }
+    return assertions;
+  }
+
+  static List<AttributeAssertion> batchReceiveAttributes(
+      String destination, String messageId, boolean testHeaders) {
+    return receiveAttributes(destination, messageId, testHeaders, true);
+  }
+
+  static List<AttributeAssertion> receiveAttributes(
+      String destination, String messageId, boolean testHeaders) {
+    return receiveAttributes(destination, messageId, testHeaders, false);
+  }
+
+  static List<AttributeAssertion> receiveAttributes(
+      String destination, String messageId, boolean testHeaders, boolean isBatch) {
+    List<AttributeAssertion> assertions =
+        new ArrayList<>(
+            Arrays.asList(
+                equalTo(SemanticAttributes.MESSAGING_SYSTEM, "pulsar"),
+                equalTo(SemanticAttributes.SERVER_ADDRESS, brokerHost),
+                equalTo(SemanticAttributes.SERVER_PORT, brokerPort),
+                equalTo(SemanticAttributes.MESSAGING_DESTINATION_NAME, destination),
+                equalTo(SemanticAttributes.MESSAGING_OPERATION, "receive"),
+                equalTo(SemanticAttributes.MESSAGING_MESSAGE_ID, messageId),
+                satisfies(
+                    SemanticAttributes.MESSAGING_MESSAGE_BODY_SIZE,
+                    AbstractLongAssert::isNotNegative)));
+    if (testHeaders) {
+      assertions.add(
+          equalTo(
+              AttributeKey.stringArrayKey("messaging.header.test_message_header"),
+              Collections.singletonList("test")));
+    }
+    if (isBatch) {
+      assertions.add(
+          satisfies(
+              SemanticAttributes.MESSAGING_BATCH_MESSAGE_COUNT, AbstractLongAssert::isPositive));
+    }
+    return assertions;
+  }
+
+  static List<AttributeAssertion> processAttributes(
+      String destination, String messageId, boolean testHeaders) {
+    List<AttributeAssertion> assertions =
+        new ArrayList<>(
+            Arrays.asList(
+                equalTo(SemanticAttributes.MESSAGING_SYSTEM, "pulsar"),
+                equalTo(SemanticAttributes.MESSAGING_DESTINATION_NAME, destination),
+                equalTo(SemanticAttributes.MESSAGING_OPERATION, "process"),
+                equalTo(SemanticAttributes.MESSAGING_MESSAGE_ID, messageId),
+                satisfies(
+                    SemanticAttributes.MESSAGING_MESSAGE_BODY_SIZE,
+                    AbstractLongAssert::isNotNegative)));
+    if (testHeaders) {
+      assertions.add(
+          equalTo(
+              AttributeKey.stringArrayKey("messaging.header.test_message_header"),
+              Collections.singletonList("test")));
+    }
+    return assertions;
+  }
+
+  static void acknowledgeMessage(Consumer<String> consumer, Message<String> message) {
+    try {
+      consumer.acknowledge(message);
+    } catch (PulsarClientException exception) {
+      throw new RuntimeException(exception);
+    }
+  }
+
+  static void acknowledgeMessages(Consumer<String> consumer, Messages<String> messages) {
```

Comment:
This is duplicated with the method right above it.

Reply:
There is a small difference. One takes `Message<String>` and the other takes `Messages<String>`